gno.land/pkg/gnoland/app.go

```diff
@@ -108,14 +97,28 @@ func NewAppWithOptions(cfg *AppOptions) (abci.Application, error) {
 			newCtx sdk.Context, res sdk.Result, abort bool,
 		) {
 			// Override auth params.
-			ctx = ctx.WithValue(
-				auth.AuthParamsContextKey{}, auth.DefaultParams())
+			ctx = ctx.
+				WithValue(auth.AuthParamsContextKey{}, auth.DefaultParams())
 			// Continue on with default auth ante handler.
 			newCtx, res, abort = authAnteHandler(ctx, tx, simulate)
 			return
 		},
 	)
 
+	// Set begin and end transaction hooks.
```

Comment:
We need the calls to create a gno transactional store, and then to commit its result, to happen precisely in the same way as when we create a "cache-wrapping" on the database stores and write to them.

https://github.com/gnolang/gno/blob/40fae4827aef89e22d740f570a64054376f95cb5/tm2/pkg/sdk/baseapp.go#L823-L846

The only other way that I can see we can do this, and which I previously attempted in a previous version of this PR, is to wrap the gno Store so that it is initialized together with other stores (iavlStore and baseStore): see commit 57f6f47c0895061faa21ddbc6ef1c8009916a1b3.

However, this is already trying to shove functionality into a `Store` interface which in its implementation is not meant to support other non-key-value application stores like this. You can see how this worked in the commit I mentioned; the problem that was popping up, was that when we use the stores in ABCI queries, they are wrapped within another store implementation inside tm2; so it's not possible for us to assert the store into a `gnostore.Store` and get the resulting gnolang.Store from there.

So, at that point, it was a choice between "do we add a `Context()` method (or similar mechanism) to the Store interface, to allow the application setting arbitrary data" or adding the hooks here. Alas, this one seems to me like a preferable solution - and also one that can be justified as useful in isolation (I think it's useful to have).

Note that we cannot have this store-committing functionality just within the VM keeper, because as you can see in the above snippet, if one message fails, the entire transaction fails and also previous succeeding messages are rolled back.

Reply:
An alternative here is to rollout our own BaseApp